src/core/unit-varlink.c

```diff
@@ -0,0 +1,234 @@
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("UpheldBy", unit_dependencies_build_json, u),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("ConflictedBy", unit_dependencies_build_json, u),
+
+                        /* Others */
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("Triggers", unit_dependencies_build_json, u),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("TriggeredBy", unit_dependencies_build_json, u),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("AccessSELinuxContext", u->access_selinux_context),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("FragmentPath", u->fragment_path),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("SourcePath", u->source_path),
+                        JSON_BUILD_PAIR_STRV_NON_EMPTY("DropInPaths", u->dropin_paths),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("UnitFilePreset", preset_action_past_tense_to_string(unit_get_unit_file_preset(u))),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("Transient", u->transient),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("Perpetual", u->perpetual));
+
+        // TODO follow up PRs:
+        // JSON_BUILD_PAIR_CALLBACK_NON_NULL("CGroup", cgroup_context_build_json, u)
+        // JSON_BUILD_PAIR_CALLBACK_NON_NULL("Exec", exec_context_build_json, u)
+        // JSON_BUILD_PAIR_CALLBACK_NON_NULL("Kill", kill_context_build_json, u)
+        // Mount/Automount context
+        // Path context
+        // Scope context
+        // Swap context
+        // Timer context
+        // Service context
+        // Socket context
+}
+
+static int list_unit_one(sd_varlink *link, Unit *unit, bool more) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(unit);
+
+        r = sd_json_buildo(&v, SD_JSON_BUILD_PAIR_CALLBACK("context", unit_context_build_json, unit));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+int vl_method_list_units(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        Manager *m = ASSERT_PTR(userdata);
+        int r;
+
+        assert(parameters);
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        const char *k;
+        Unit *u, *previous = NULL;
+        HASHMAP_FOREACH_KEY(u, k, m->units) {
+                if (k != u->id)
+                        continue;
+
+                /* TODO: we need some kind of filtering to show only necessary units.
+                 * However, I'm leaving this to a follow up commits to reduce amount of changes
+                 * introduced in one go.
+                 * if (unit_is_filtered(u, p.states, p.patterns)) */
+                        /* continue; */
```

Comment:
Hmm can you drop this comment actually, we prefer "perfect" commits in systemd, so adding stuff here that is removed in a later commit is frowned upon.